README.md
# keen-js-simple

A simple, super lightweight (3kb) & performance optimized library to track data with [Keen.IO](https://keen.io).

[![Travis build status](http://img.shields.io/travis/mydea/keen-js-simple.svg?style=flat)](https://travis-ci.org/mydea/keen-js-simple)

## Difference to keen-js

Keen.IO provides a full-featured JavaScript SDK, [keen-js](https://github.com/keen/keen-js), which can be used to track and query data from the Keen.IO API. However, this has two main disadvantages:

1. Because it is that full-featured, it is quite heavy weight (34kb minified for the tracking-only version).
2. It sends one request for every event that is tracked. This can lead to many simultaneous Ajax requests, which is bad for performance.

This small library provides a simple way to send events to Keen.IO, which will automatically be debounced and combined to reduce the number of HTTP requests. 

## Installation

You can install keen-js-simple via bower (or npm):

```bash
bower install keen-js-simple --save
```

Then, you can simply include it in your HTML:

```html
<script src="bower_components/keen-js-simple/dist/keen.min.js"></script>
```

Or, if you are using modules, you can include it:

```js
import keen from './../bower_components/keen-js-simple/dist/keen.min.js';
```

For Ember.js, see [ember-keen](https://github.com/mydea/ember-keen) on which this library is based.

## Usage

In the simplest form, you will need to include the script, and then do minimal configuration:

```html
<script>
  keen.setOptions({
    projectId: 'MY-PROJECT-ID',
    writeKey: 'MY-WRITE-KEY'
  });
</script>
```

Then, you'll be able to send events like this:

```js
keen.sendEvent('simple-event');
keen.sendEvent('event-name', {
  property1: 1
});
```

These two events will then be sent in one payload.
By default, it will wait for 5 seconds until sending data. This can be configured via `keen.setOptions({ queueTime: 4000 })`.

By default, this will try to send the data via jQuery's `$.post()` method. 
However, this can easily be overridden in the options (see below), removing the dependency on jQuery.

## Configuration

There are a few other configuration options. Note that you can set one or more options at the same time with `keen.setOptions()`.

```js
keen.setOptions({
  projectId: null,
  writeKey: null,
  queueTime: 5000, // Time in ms to wait before sending
  baseURL: 'https://api.keen.io/3.0/projects',
  
  post: function(data, options) {
    // Actually send the (combined) data to the API
    // By default, this uses $.post
  },
  
  mergeData: function(data, options) {
    // This should return an object which will be merged with the data to send
    // This can be used to add things like the user agent to all events
    // By default, we set the timestamp property to now, 
    // otherwise the timestamps will be set to the time of the API request
  
    return {
      keen: {
        timestamp: new Date()
      }
    };
  }
});
```

You can set the `post` option if you do not want to use jQuery to make the Ajax request. The default implementation is:

```js
keen.setOptions({
  post: function(data, options) {
    var baseUrl = options.baseURL;
    var projectId = options.projectId;
    var writeKey = options.writeKey;
     
    var url = baseUrl + '/' + projectId + '/events?api_key=' + writeKey;
     
    return $.ajax({
      type: 'POST',
      headers: {
       Authorization: writeKey
      },
      url: url,
      contentType: 'application/json',
      crossDomain: true,
      xhrFields: {
       withCredentials: false
      },
      data: JSON.stringify(data),
      dataType: 'json'
    });
  }
});
```

The `mergeData` function can be overridden to append global data to every event. For example, you might want to add user agent information to all events:

```js
keen.setOptions({
  mergeData: function(data, options) {
      return {
        keen: {
          timestamp: new Date()
        },
        userAgent: window.navigator.userAgent
      };
    }
})
```

It is recommended to always include the `keen.timestamp` portion, as otherwise the timestamps in keen will be from the time when the events are received by the API, not when they where triggered on the client.

## Methods

There are only a few methods available on the keen-class:

```js
keen.setOptions({}); // See configuration section
keen.sendEvent('event-name', {}); // Data is optional
keen.getQueue(); // Get all queued events
keen.clearQueue(); // Remove all queued events
```

## Changelog

For changes, see the [Changelog](CHANGELOG.md)

## Copyright
Copyright © 2017 Francesco Novy | MIT license


CHANGELOG.md
### [0.1.1](https://github.com/mydea/keen-js-simple/releases/tag/0.1.1)

- The first release


test/unit/keen-js-simple.js
import globalKeen from '../../src/keen-js-simple';

describe('keen', () => {
  describe('Options work', () => {

    it('should have the correct default options', () => {
      let keen = globalKeen;
      let options = keen.options;

      expect(typeof options).to.equal('object');
      expect(options.projectId).to.equal(null);
      expect(options.writeKey).to.equal(null);
      expect(options.queueTime).to.equal(5000);
      expect(options.baseURL).to.equal('https://api.keen.io/3.0/projects');
    });

    it('should allow changing of single options', () => {
      let keen = globalKeen.createInstance();
      let options = keen.options;

      expect(typeof options).to.equal('object');
      expect(options.projectId).to.equal(null);
      expect(options.writeKey).to.equal(null);
      expect(options.queueTime).to.equal(5000);
      expect(options.baseURL).to.equal('https://api.keen.io/3.0/projects');

      keen.setOptions({
        projectId: 'test1'
      });

      expect(keen.options.projectId).to.equal('test1');
      expect(keen.options.writeKey).to.equal(null);

      keen.setOptions({
        writeKey: 'test2'
      });

      expect(keen.options.projectId).to.equal('test1');
      expect(keen.options.writeKey).to.equal('test2');
    });

    it('should allow chaning of multiple options', () => {
      let keen = globalKeen.createInstance();

      let postFunc = function() {
      };

      let mergeDataFunc = function() {
      };

      keen.setOptions({
        projectId: 'test1',
        writeKey: 'test2',
        queueTime: 3000,
        baseURL: 'test3',
        post: postFunc,
        mergeData: mergeDataFunc
      });

      let options = keen.options;

      expect(typeof options).to.equal('object');
      expect(options.projectId).to.equal('test1');
      expect(options.writeKey).to.equal('test2');
      expect(options.queueTime).to.equal(3000);
      expect(options.baseURL).to.equal('test3');
      expect(options.post).to.equal(postFunc);
      expect(options.mergeData).to.equal(mergeDataFunc);
    });

  });
});


src/keen-js-simple.js
/**
 * This library allows sending data to Keen.IO.
 *
 * @class Keen
 * @public
 */
const Keen = function() {

  /**
   * Set global options.
   * Only set options will be set, so you can safely set only single options like this:
   *
   * ´´´js
   * keen.setOptions({ queueTime: 4000 });
   * ```
   *
   * @method setOptions
   * @param {Object} newOptions The new options to set
   * @return {Object} The full options object
   * @public
   */
  let setOptions = function(newOptions) {
    if (typeof newOptions !== 'object') {
      return;
    }

    if (newOptions.hasOwnProperty('projectId')) {
      options.projectId = newOptions.projectId;
    }

    if (newOptions.hasOwnProperty('writeKey')) {
      options.writeKey = newOptions.writeKey;
    }

    if (newOptions.hasOwnProperty('post')) {
      if (typeof newOptions.post !== 'function') {
        throw new Error('post needs to be a function!');
      }
      options.post = newOptions.post;
    }

    if (newOptions.hasOwnProperty('baseURL')) {
      options.baseURL = newOptions.baseURL;
    }

    if (newOptions.hasOwnProperty('queueTime')) {
      if (typeof newOptions.queueTime !== 'number') {
        throw new Error('queueTime needs to be a number, representing the ms to wait for other events!');
      }
      options.queueTime = newOptions.queueTime;
    }

    if (newOptions.hasOwnProperty('mergeData')) {
      if (typeof newOptions.mergeData !== 'function') {
        throw new Error('mergeData needs to be a function!');
      }
      options.mergeData = newOptions.mergeData;
    }

    return options;
  };

  /**
   * Send an event to Keen.
   * This will be debounced for the number of ms defined in queueTime.
   *
   * @method sendEvent
   * @param {String} event The name of the event to send
   * @param {Object} plainData The data to attach to the event.
   * @public
   */
  let sendEvent = function(event, plainData) {
    if (!plainData) {
      plainData = {};
    }
    if (typeof plainData !== 'object') {
      throw new Error('data must be an object');
    }

    let data = _prepareEventData(plainData);

    let queue = _eventQueue;
    if (queue.hasOwnProperty(event)) {
      queue[event].push(data);
    } else {
      queue[event] = [data];
    }

    _debounceTimer();
  };

  /**
   * Get the queued events.
   *
   * @method getQueue
   * @return {Object}
   * @public
   */
  let getQueue = function() {
    return _eventQueue;
  };

  /**
   * Clear all queued events.
   *
   * @method clearQueu
   * @public
   */
  let clearQueue = function() {
    _eventQueue = {};
  };

  /**
   * Create a new keen instance.
   *
   * @method createInstance
   * @return {Keen}
   * @public
   */
  let createInstance = function() {
    return new Keen();
  };

  /**
   * The default post function uses jQuery's $.post to send data.
   *
   * @method _post
   * @param {Object} data The full data to send
   * @param {Object} options The options are passed and can be used to build the ajax request
   * @return {jQuery.post}
   * @private
   */
  let _post = function(data, options) {
    let baseUrl = options.baseURL;
    let projectId = options.projectId;
    let writeKey = options.writeKey;

    // Check if all dependencies are loaded && all properties are set
    if (typeof $ === 'undefined') {
      throw new Error('The default post method of keen requires jQuery. You need to either include jQuery, or provide a custom post method.');
    }

    if (!baseUrl) {
      throw new Error('The baseURL of keen is not set correctly.');
    }

    if (!projectId) {
      throw new Error('The projectId of keen is not set correctly.');
    }

    if (!writeKey) {
      throw new Error('The writeKey of keen is not set correctly.');
    }

    let url = baseUrl + '/' + projectId + '/events?api_key=' + writeKey;

    return $.ajax({
      type: 'POST',
      headers: {
        Authorization: writeKey
      },
      url: url,
      contentType: 'application/json',
      crossDomain: true,
      xhrFields: {
        withCredentials: false
      },
      data: JSON.stringify(data),
      dataType: 'json'
    });
  };

  /**
   * This function merges the data from mergeData() into event data.
   *
   * @method _prepareEventData
   * @param {Object} data The object to merge data into. This will be mutated!
   * @return {Object} The object with the data merged into it
   * @private
   */
  let _prepareEventData = function(data) {
    let mergeData = options.mergeData();

    for (let i in mergeData) {
      if (mergeData.hasOwnProperty(i)) {
        data[i] = mergeData[i];
      }
    }

    return data;
  };

  /**
   * Debounce the timer to process the queue.
   * This will wait for queueTime ms to actually run the queue.
   * If the function is called again before that time, it will be reset.
   * This means that the queue runs only if no data has been added for the specified time.
   *
   * @method _debounceTimer
   * @private
   */
  let _debounceTimer = function() {
    let queueTime = options.queueTime;
    let timer = _queueTimer;

    if (timer) {
      window.clearTimeout(timer);
    }

    _queueTimer = window.setTimeout(_processQueue, queueTime);
  };

  /**
   * Actually process the queue and send the data.
   *
   * @method _processQueu
   * @private
   */
  let _processQueue = function() {
    let data = _eventQueue;
    options.post(data, options);
    _eventQueue = {};
  };

  /**
   * The keen options.
   * These can be set via keen.setOptions().
   *
   * @property options
   * @type {Object}
   * @protected
   */
  let options = {
    projectId: null,
    writeKey: null,
    queueTime: 5000,
    baseURL: 'https://api.keen.io/3.0/projects',
    post: _post,
    mergeData: function() {
      return {
        keen: {
          timestamp: new Date()
        }
      };
    }
  };

  /**
   * This hols the queued events, keyed by event name.
   *
   * @property _eventQueue
   * @type {Object}
   * @private
   */
  let _eventQueue = {};

  /**
   * This is the reference to the queue timer, which is used so it can be canceled.
   *
   * @property _queueTimer
   * @type {null|Timeout}
   * @private
   */
  let _queueTimer = null;

  return {
    setOptions,
    options,
    sendEvent,
    getQueue,
    clearQueue,
    createInstance
  };
};

const keen = new Keen();

export default keen;
